storey/sources.py

```diff
@@ -622,49 +621,78 @@ def __init__(
         self._dfs = dfs
         self._key_field = key_field
         self._id_field = id_field
 
+    def _get_key(self, body):
+        key = None
+        if self._key_field:
+            if isinstance(self._key_field, list):
+                key = []
+                for key_field in self._key_field:
+                    single_key = self._get_by_field_or_index(field=key_field, body=body)
+                    if pd.isna(single_key):
+                        return single_key, key_field
+                    key.append(single_key)
+
+            else:
+                key = self._get_by_field_or_index(field=self._key_field, body=body)
+                if pd.isna(key):
+                    return key, self._key_field
+        return key, None
+
+    def _get_element(self, body: OrderedDict):
+        return dict(body)
+
     async def _run_loop(self):
         for df in self._dfs:
             for namedtuple in df.itertuples():
-                create_event = True
-                body = namedtuple._asdict()
+                body = OrderedDict(namedtuple._asdict())
                 index = body.pop("Index")
                 if len(df.index.names) > 1:
                     for i, index_column in enumerate(df.index.names):
                         body[index_column] = index[i]
                 elif df.index.names[0] is not None:
                     body[df.index.names[0]] = index
-
-                key = None
-                if self._key_field:
-                    if isinstance(self._key_field, list):
-                        key = []
-                        for key_field in self._key_field:
-                            if key_field not in body or pandas.isna(body[key_field]):
-                                create_event = False
-                                break
-                            key.append(body[key_field])
-                    else:
-                        key = body[self._key_field]
-                        if key is None:
-                            create_event = False
-                if create_event:
+                key, none_key = self._get_key(body=body)
+                if not none_key:
                     if self._id_field:
-                        id = body[self._id_field]
+                        line_id = self._get_by_field_or_index(field=self._id_field, body=body)
                     else:
-                        id = self._get_uuid()
-                    event = Event(body, key=key, id=id)
+                        line_id = self._get_uuid()
+                    element = self._get_element(body=body)
+                    event = Event(element, key=key, id=line_id)
                     await self._do_downstream(event)
                 else:
                     if self.context:
-                        self.context.logger.error(f"For {body} value of key {key_field} is None")
+                        self.context.logger.error(f"value of key {none_key} is None For {body}")
```

Comment:
```suggestion
                        self.context.logger.error(f"value of key column '{none_key}' is None for {body}")
```
better yet, replace None with the actual value, which may be NaN rather than None.